### PR TITLE
Point cloud shader

### PR DESCRIPTION
**Pull request by Vesper-Works**

### CODE REVIEW

File: Assets/Scripts/Point Cloud/PointCloudObject.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+            Debug.Log("vertex Count :" + _vertices.Count);
+
+            _pointBuffer = new ComputeBuffer(_vertices.Count, 3 * sizeof(float));
+            _pointBuffer.SetData(_vertices);
+            //computeShader.SetBuffer(0, "_Points", pointBuffer);
+        }
+        private void GetNormalsDataFromMesh()
+        {
+            foreach (Mesh mesh in _meshes)
```
Comment (kit-calloway) [inline Assets/Scripts/Point Cloud/PointCloudObject.cs]: This `foreach` loop doesn't need braces - it has only 1 line

File: Assets/Scripts/Point Cloud/PointCloudObject.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            {
+                _normals.AddRange(mesh.normals);
+            }
+
+            Debug.Log("normals Count :" + _normals.Count);
+
+            _normalBuffer = new ComputeBuffer(_normals.Count, 3 * sizeof(float));
+            _normalBuffer.SetData(_normals);
+            //computeShader.SetBuffer(0, "_Points", pointBuffer);
```
Comment (kit-calloway) [inline Assets/Scripts/Point Cloud/PointCloudObject.cs]: This comment should be removed.

File: Assets/Scripts/Point Cloud/PointCloudObject.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            foreach (Mesh mesh in _meshes)
+            {
+                _vertices.AddRange(mesh.vertices);
+            }
+
+            Debug.Log("vertex Count :" + _vertices.Count);
+
+            _pointBuffer = new ComputeBuffer(_vertices.Count, 3 * sizeof(float));
+            _pointBuffer.SetData(_vertices);
+            //computeShader.SetBuffer(0, "_Points", pointBuffer);
```
Comment (kit-calloway) [inline Assets/Scripts/Point Cloud/PointCloudObject.cs]: This comment should be removed.

File: Assets/Scripts/Point Cloud/PointCloudObject.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        {
+            if (parent.TryGetComponent(out MeshFilter meshFilter))
+            {
+                _meshes.Add(meshFilter.mesh);
+            }
+        }
+
+        private void GetPositionsDataFromMesh()
+        {
+            foreach (Mesh mesh in _meshes)
```
Comment (kit-calloway) [inline Assets/Scripts/Point Cloud/PointCloudObject.cs]: This `foreach` loop doesn't need braces - it has only 1 line

File: Assets/Scripts/Point Cloud/PointCloudObject.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        private void Update()
+        {
+            Graphics.DrawMeshInstancedProcedural(pointMesh, 0, pointMaterial, _bounds, _pointBuffer.count);
+        }
+
+        private void GetEachMeshRecursively(Transform parent)
+        {
+            if (parent.TryGetComponent(out MeshFilter meshFilter))
```
Comment (kit-calloway) [inline Assets/Scripts/Point Cloud/PointCloudObject.cs]: Remove braces

File: Assets/Scripts/Point Cloud/PointCloudObject.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            GetPositionsDataFromMesh();
+            GetNormalsDataFromMesh();
+            pointMaterial.SetBuffer(PositionsId, _pointBuffer);
+            pointMaterial.SetBuffer("_Normals", _normalBuffer);
+            pointMaterial.SetFloat(StepId, _step);
+            pointMaterial.SetFloat("_scale", this.transform.localScale.x);
+            pointMaterial.SetFloat("_intensity", 1);
+            pointMaterial.SetVector("_worldPos", this.transform.position);
+            pointMaterial.SetVector("_color", new Vector4(1, 1, 1, 1));
+            //material.SetFloat("_ColorFromTextureLerp", colorFromTextureLerp);
```
Comment (kit-calloway) [inline Assets/Scripts/Point Cloud/PointCloudObject.cs]: These comments should be removed

File: Assets/Scripts/Point Cloud/PointCloudObject.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            //material.SetFloat("_UseAlpha", useAlpha?1:0);
+            pointMaterial.SetMatrix("_quaternion",
+                Matrix4x4.TRS(new Vector3(0, 0, 0), transform.rotation, new Vector3(1, 1, 1)));
```
Comment (kit-calloway) [inline Assets/Scripts/Point Cloud/PointCloudObject.cs]: Inconsistent indentation. To provide context, I would extract variables for the first and last arguments in `Matrix4x4.RTS()`, as well as the function return itself.

File: Assets/Scripts/Point Cloud/PointCloudObject.cs [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+using System.Linq;
+using UnityEngine.Serialization;
+
+namespace cmp2804.Point_Cloud
+{
+    public class PointCloudObject : MonoBehaviour
+    {
+        [SerializeField] private Material pointMaterial = default;
+        [SerializeField] private Mesh pointMesh = default;
+        [SerializeField] private ComputeShader computeShader;
```
Comment (kit-calloway) [inline Assets/Scripts/Point Cloud/PointCloudObject.cs]: For consistency (and functionality), replace all `[SerializeField]` with `[OdinSerialize]` as this supports more advanced types

File: Assets/Scripts/Point Cloud/PointCloudObject.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
+using System;
+using UnityEngine;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine.Serialization;
+
+namespace cmp2804.Point_Cloud
+{
+    public class PointCloudObject : MonoBehaviour
```
Comment (kit-calloway) [inline Assets/Scripts/Point Cloud/PointCloudObject.cs]: Derive from `SerializedMonoBehaviour` instead for access to property serialization

File: Assets/Scripts/Point Cloud/PointCloudRenderer.cs [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+            pointMaterial.SetFloat(StepId, PointScale);
+            pointMaterial.SetFloat("scale", transform.localScale.x);
+            pointMaterial.SetFloat("intensity", 1);
+            pointMaterial.SetVector("worldPos", transform.position);
+            pointMaterial.SetMatrix("quaternion",
+                Matrix4x4.TRS(new Vector3(0, 0, 0), transform.rotation, new Vector3(1, 1, 1)));
+        }
+        private void Update()
+        {
+            // for (int i = 0; i < _lifespans.Count; i++)
```
Comment (kit-calloway) [inline Assets/Scripts/Point Cloud/PointCloudRenderer.cs]: Pleass remove comments

File: Assets/Scripts/Point Cloud/PointCloudRenderer.cs [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+            computeShader.SetBuffer(0, "lifespanScales", _lifespanScaleBuffer);
+            computeShader.Dispatch(0, Mathf.Max(1, Mathf.CeilToInt(_points.Count / 32f)), 1, 1);
+            float[] newLifespanData = new float[_lifespans.Count];
+
+            _lifespanBuffer.GetData(newLifespanData);
+            _lifespans = newLifespanData.ToList();
+            bool reset = false;
+            for (int i = 0; i < _lifespans.Count; i++)
+            {
+                if (_lifespans[i] <= 0)
```
Comment (kit-calloway) [inline Assets/Scripts/Point Cloud/PointCloudRenderer.cs:71]: I would recommend extracting the contents of this `if` statement and replacing it with `if (_lifespans[i] <= 0) functionName();`

File: Assets/Scripts/Point Cloud/SoundMaker.cs [hunk 12]
```diff
@@ -0,0 +1,88 @@
+using System;
+using System.Collections;
+using Sirenix.OdinInspector;
+using UnityEngine;
+
+namespace cmp2804.Point_Cloud
+{
+    public class SoundMaker : MonoBehaviour
+    {
+        [Header("Emission settings")] [SerializeField]
```
Comment (kit-calloway) [inline Assets/Scripts/Point Cloud/SoundMaker.cs]: Recommend replacing this with `[Title(title, description)` as this is Odin's implementation

**Review by StevieGoodman — CHANGES_REQUESTED** [review]
Okay, there's nothing game-breaking here, but I've given a few tips that I would recommend acting upon, but don't feel compelled to

**Review by StevieGoodman — CHANGES_REQUESTED** [review]
Please resolve merge conflicts

**Comment by Vesper-Works** [comment]
Ready

**Review by StevieGoodman — APPROVED** [review]
https://media.istockphoto.com/id/513921039/photo/illustration-of-a-oool-yellow-smiley-with-sunglasses.jpg?s=612x612&w=0&k=20&c=hhVQxXTUhmcZLv2QrZ2WE2p7inzxQIA5H6XP8jPrQXw=

**Pull request merged**